File: rust/packages/blockchain/src/mempool.ts
import type { Transaction } from '@noetis/protocol';
import { randomUUID } from 'node:crypto';

export class Mempool {
  private txs = new Map<string, Transaction>();
  private seen = new Set<string>();

  add(tx: Transaction): boolean {
    const key = tx.id || `${tx.type}:${tx.from}:${tx.timestamp}`;
    if (this.seen.has(key)) return false;
    this.seen.add(key);
    this.txs.set(tx.id, tx);
    return true;
  }

  list(): Transaction[] {
    return [...this.txs.values()].sort((a, b) => a.timestamp - b.timestamp);
  }

  drain(): Transaction[] {
    const all = this.list();
    this.txs.clear();
    return all;
  }

  remove(ids: string[]): void {
    for (const id of ids) this.txs.delete(id);
  }

  size(): number {
    return this.txs.size;
  }
}

export function createTx(
  type: Transaction['type'],
  from: string | null,
  to: string | null,
  amount: number,
  metadata: Record<string, unknown> = {}
): Transaction {
  return {
    id: randomUUID(),
    type,
    from,
    to,
    amount,
    metadata,
    timestamp: Date.now(),
  };
}

File: rust/crates/noetis-blockchain/src/mempool.rs
use noetis_protocol::{Transaction, TransactionType};
use serde_json::Value;
use std::collections::{HashMap, HashSet};
use uuid::Uuid;

pub struct Mempool {
    txs: HashMap<String, Transaction>,
    seen: HashSet<String>,
}

impl Default for Mempool {
    fn default() -> Self {
        Self::new()
    }
}

impl Mempool {
    pub fn new() -> Self {
        Self {
            txs: HashMap::new(),
            seen: HashSet::new(),
        }
    }

    pub fn add(&mut self, tx: Transaction) -> bool {
        let key = if tx.id.is_empty() {
            format!(
                "{:?}:{}:{}",
                tx.tx_type,
                tx.from.as_deref().unwrap_or(""),
                tx.timestamp
            )
        } else {
            tx.id.clone()
        };
        if self.seen.contains(&key) {
            return false;
        }
        self.seen.insert(key);
        self.txs.insert(tx.id.clone(), tx);
        true
    }

    pub fn list(&self) -> Vec<Transaction> {
        let mut txs: Vec<Transaction> = self.txs.values().cloned().collect();
        txs.sort_by_key(|t| t.timestamp);
        txs
    }

    pub fn drain(&mut self) -> Vec<Transaction> {
        let all = self.list();
        self.txs.clear();
        all
    }

    pub fn remove(&mut self, ids: &[String]) {
        for id in ids {
            self.txs.remove(id);
        }
    }

    pub fn size(&self) -> usize {
        self.txs.len()
    }
}

pub fn create_tx(
    tx_type: TransactionType,
    from: Option<&str>,
    to: Option<&str>,
    amount: f64,
    metadata: HashMap<String, Value>,
) -> Transaction {
    Transaction {
        id: Uuid::new_v4().to_string(),
        tx_type,
        from: from.map(str::to_string),
        to: to.map(str::to_string),
        amount,
        metadata,
        timestamp: std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_millis() as i64,
        signature: None,
    }
}
